### PR TITLE
Embed metadata about the current page instead of parsing current path

### PR DESCRIPTION
**Pull request by Nemo157**
To help out with the client-side for #2434

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Makes the route configuration and the code much simpler. Good idea!

**Comment by Nemo157** [comment]
When working out how to test the changes I realised there's even a much simpler way to handle the menus, that let the tests verify the ajax menu even better.

File: static/index.js [hunk 1]
```diff
@@ -14,23 +9,9 @@ function formatCrateName(crateTitleAndVersion) {
...
-
-            document.body.append(temporaryInput);
-            temporaryInput.select();
-            document.execCommand("copy");
-
-            temporaryInput.remove();
-            crateTitleAndVersion.style.textTransform = oldTransform;
+        async function copyTextHandler() {
+            const metadata = JSON.parse(document.getElementById("crate-metadata").innerText)
+            await navigator.clipboard.writeText(`${metadata.name} = "${metadata.version}"`)
```
Comment (Nemo157) [inline static/index.js:14]: Ah, one thing I should note is that `navigator.clipboard` requires a secure context, so this won't work when running on an http-only dev instance (I have my dev instance behind a reverse proxy adding https to it so was able to verify it works, on firefox at least).

Comment (GuillaumeGomez) [reply static/index.js:14]: So, you can actually cheat here to do that (we do it in rustdoc for example):

```js
const el = document.createElement("textarea");
el.value = `${metadata.name} = "${metadata.version}"`;
el.setAttribute("readonly", "");
// To not make it appear on the screen.
el.style.position = "absolute";
el.style.left = "-9999px";

document.body.appendChild(el);
el.select();
document.execCommand("copy");
document.body.removeChild(el);
```

No need for https anymore. :)

Comment (GuillaumeGomez) [reply static/index.js:14]: Oh that's kinda what we were doing before in fact.

File: static/menu.js [hunk 2]
```diff
@@ -7,42 +7,27 @@ const updateMenuPositionForSubMenu = (currentMenuSupplier) => {
...
         return;
     }
-    const parts = window.location.pathname.split("/");
-    let crateName = parts[1];
-    let version = parts[2];
-    if (crateName === "crate") {
-        crateName = parts[2];
-        version = parts[3];
-        path += "/crate";
+    let url = listElem.dataset.url
```
Comment (GuillaumeGomez) [inline static/menu.js:23]:
```suggestion
    const url = listElem.dataset.url;
```

File: static/index.js (hunk 1 above)
Comment (Nemo157) [reply static/index.js:14]: Yeah, I sort of wanted to get away from that unspecced API to the official replacement, and for docs.rs I don't think the insecure-context usecase is very important.

Comment (GuillaumeGomez) [reply static/index.js:14]: But for testing locally, it is. I think we should keep the "unspecced API" (which in fact is, it's just a round-about way to do the same) for making it easier to test changes locally.

Comment (Nemo157) [reply static/index.js]: Reverted this part of the change.

**Comment by syphar** [comment]
@Nemo157 is this waiting for review again?

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
@syphar Would be better to have yours. ;)

**Comment by Nemo157** [comment]
Yep, it should be ready, IIRC I just fixed the two issues @GuillaumeGomez brought up.

**Pull request merged**